CMakeLists.txt

```diff
@@ -0,0 +1,630 @@
+if(WIN32)
+    set(CMAKE_SHARED_LIBRARY_PREFIX "")
+endif()
+# -----------------------------------
+
+
+if(POLICY CMP0077) # Cache variables override since 3.12
+    cmake_policy(SET CMP0077 NEW)
+endif()
+
+if(EMSCRIPTEN)
+    set(BUILD_SHARED_DEFAULT OFF)
+else()
+    set(BUILD_SHARED_DEFAULT ON)
+endif()
+
+option(BUILD_STATIC   "Build static library of libXMP" ON)
+option(BUILD_SHARED   "Build shared library of libXMP" ${BUILD_SHARED_DEFAULT})
+if(NOT BUILD_STATIC AND NOT BUILD_SHARED)
+    message(FATAL_ERROR "Both static and shared builds got disabled. You must enable at least one of them.")
+endif()
+
+option(LIBXMP_BUILD_LITE   "Build lite version of libXMP" ON)
+option(LIBXMP_BUILD_FULL   "Build full version of libXMP" ON)
+if(NOT LIBXMP_BUILD_LITE AND NOT LIBXMP_BUILD_FULL)
+    message(FATAL_ERROR "Both Full and Lite versions got disabled. You must enable at least one of them.")
+endif()
+
+option(LIBXMP_DISABLE_IT   "Disable IT support at libXMP-Lite" OFF)
+option(LIBXMP_DISABLE_DEPACKERS     "Disable archive depackers at libXMP" OFF)
+option(LIBXMP_DISABLE_PROWIZARD     "Disable the ProWizard at libXMP" OFF)
+
+TEST_BIG_ENDIAN(WORDS_BIGENDIAN)
+if(WORDS_BIGENDIAN)
+    add_definitions(-DWORDS_BIGENDIAN=1)
+endif()
+
+if(CMAKE_BUILD_TYPE_LOWER STREQUAL "debug")
+    add_definitions(-DDEBUG -D_DEBUG -DDEBUG_NOCOLOR)
+endif()
+
+check_include_file(alloca.h HAVE_ALLOCA_H)
+if(HAVE_ALLOCA_H)
+    add_definitions(-DHAVE_ALLOCA_H=1)
+endif()
+
+check_symbol_exists(round "math.h;tgmath.h" HAVE_ROUND_FUNCTION)
+if(HAVE_ROUND_FUNCTION)
+    add_definitions(-DHAVE_ROUND=1)
+endif()
+
+check_symbol_exists(powf "math.h;tgmath.h" HAVE_POWF_FUNCTION)
+if(HAVE_POWF_FUNCTION)
+    add_definitions(-DHAVE_POWF=1)
+endif()
+
+check_symbol_exists(popen "stdio.h" HAVE_POPEN)
+if(HAVE_POPEN)
+    add_definitions(-DHAVE_POPEN=1)
+endif()
+
+check_symbol_exists(fnmatch "fnmatch.h" HAVE_FNMATCH)
+if(HAVE_FNMATCH)
+    add_definitions(-DHAVE_FNMATCH=1)
+endif()
+
+check_symbol_exists(umask "sys/stat.h" HAVE_UMASK)
+if(HAVE_UMASK)
+    add_definitions(-DHAVE_UMASK=1)
+endif()
+
+check_symbol_exists(wait "sys/wait.h" HAVE_WAIT)
+if(HAVE_WAIT)
+    add_definitions(-DHAVE_WAIT=1)
+endif()
+
+check_symbol_exists(pipe "unistd.h" HAVE_PIPE)
+if(HAVE_PIPE)
+    add_definitions(-DHAVE_PIPE=1)
+endif()
+
+check_symbol_exists(fork "unistd.h" HAVE_FORK)
+if(HAVE_FORK)
+    add_definitions(-DHAVE_FORK=1)
+endif()
+
+check_symbol_exists(execvp "unistd.h" HAVE_EXECVP)
+if(HAVE_EXECVP)
+    add_definitions(-DHAVE_EXECVP=1)
+endif()
+
+check_symbol_exists(dup2 "unistd.h" HAVE_DUP2)
+if(HAVE_DUP2)
+    add_definitions(-DHAVE_DUP2=1)
+endif()
+
+check_symbol_exists(mkstemp "stdlib.h" HAVE_MKSTEMP)
+if(HAVE_MKSTEMP)
+    add_definitions(-DHAVE_MKSTEMP=1)
+endif()
+
+check_include_file(dirent.h HAVE_DIRENT)
+if(HAVE_DIRENT)
+    add_definitions(-DHAVE_DIRENT=1)
+endif()
+
+if(UNIX AND NOT APPLE AND NOT HAIKU AND NOT EMSCRIPTEN)
+    find_library(M_LIBRARY m)
```

Comment:
configure.ac has a list of targets either not having or not needing libm.
Please adjust accordingly. (e.g. libsamplerate cmake file has such an
example, if it makes it easier for you.)

Reply:
The libm is REQUIRED for unit test builds in a case when shared library building is disabled, otherwise, unit tests get failed with undefined reference errors.

Reply:
No: libm is _not_ required by every and all platforms: mingw doesn't need it
(its libm.a has only an empty object), darwin (i.e. mac) doesn't need it, cygwin
doesn't need it, etc..

Reply:
That is right but on Linux, I do have the failure to build the application that links libxmp without libm. Therefore, libm code doesn't require it but tries to find its presence to use it.

Reply:
As I told you, the libm is not used for libxmp itself, it's used with applications with static libxmp linked.

Even the Autotools-built test tool has libm linked:
![изображение](https://user-images.githubusercontent.com/6751442/125408181-4f177f80-e3c3-11eb-90d1-bc19625605fd.png)

Reply:
> That is right but on Linux

No, it is _not_ right on linux: linux/glibc does need libm.
That's why I'm telling you to use correct filter to skip libm
where it's not needed

Reply:
> As I told you, the libm is not used for libxmp itself,

Oh, it _is_ used by libxmp itself

>  it's used with applications with static libxmp linked.

As I said, look at configure.ac to see what I really mean

Reply:
The attempt to link the static Autotools-built libxmp with an example:
![изображение](https://user-images.githubusercontent.com/6751442/125410389-84bd6800-e3c5-11eb-8fd0-b32cee3b39ab.png)

Reply:
> As I said, look at configure.ac to see what I really mean

Yeah, I do see it's here

Reply:
Okay, that wasn't valid, because it linked to an old library copy that was laid at the system, I did with the local build (I had to modify the makefile to add the -L../lib reference)
![Снимок экрана от 2021-07-13 11-19-47](https://user-images.githubusercontent.com/6751442/125417596-cb2ac94a-b1b5-410d-8398-876688cb83a4.png)

Reply:
The makefile in examples directory expects that you are linking to
the shared library. It really is an _example_ makefile: it doesn't even
try to find where libxmp is (that `-L../lib` switch you manually added is
the proof) and makes fiddling with it to the user.

In the above case, you tried linking to static libxmp which does need
libm, and you are possibly expected to edit the makefile and add it.

Reply:
That right, I had only shown that why I added the libm code, I need it to link the static libxmp with unit tests, it doesn't make sense for the case of shared libraries.

Reply:
For god's sake, I'm _not_ saying not to add libm linkage,
What I am saying is do something that looks like this:
https://github.com/sezero/libsamplerate/blob/master/CMakeLists.txt#L57

Reply:
Done already :wink: